content/blog/2019-11-15-profile-guided-basic-block-reordering.md

```diff
@@ -0,0 +1,90 @@
+
+In this project, we used the [build-in profile generation via instrumentation tool in Clang](https://clang.llvm.org/docs/UsersManual.html#profiling-with-instrumentation) to collect profiles. 
+
+## Block Reordering Algorithm
+
+We implemeted one variant of algo2 in Profile Guided Code Positioning as one LLVM pass. The frequencies of each branch in conditional jump and switch instrucitons are used as edges weight. We also experiemnted with marking weight of unconditional branch to be infinity.
+
+The algorithm first identifies frequently executed "chains", i.e. paths of basic blocks. It starts by setting all basic blocks as individual chains. Then the edges are iterated from highest weight to lowest. For each edge, if it is connecting tail and head of two chains, the chains will be merged. We implemented this by maintaining the start of chains and the next block relationships in hashmaps. Then we iterate through all the block terminators to collect edges with weights and rank them by weights. Finally we merge the chains if and only if the edge is jumping from the end of a chain to the start of a chain.
+
+After the chains are identified, they are arranged sequentially based on weights over edges cross chains. In our implementation, we decide the next chain to put in the way that the current chain has one branch with largest weight point to the start of it. 
+
+
+## Evaluation
+
+### Initial Testing
+Initially we tested our pass on simple programs with only one conditional branch, where one branch is more frequently executed than the other. Our pass successfully rearranged the more used pass to directly after the block before branching. We also tested on more complicated branching scenarios. 
+
+### Benchmark Testing
+
+We also evaluated the pass on [PARSEC](https://parsec.cs.princeton.edu/). Due to machine and time constraints, we only managed to run streamcluster programs with profiles provided by running the simulated inputs. We compared clang with only our optimization pass against clang with no optimization. Unfortunately, the running time and branch missing rates are not significantly improved:
+
+(will add result soon)
+
+We think this is because the profiling information we collected is not enough to reflect the actual workload, since we are only using small simulated input for profiling. Clang profiling also does not provide unconditional branch frequency. Also the branch missing rate is already pretty low even without any optimization. 
+
+### Extensions and Improvement
+
+Ideally we would like to implement more complicated position ordering and have more rigorous testing set up, including collecting more profiles and find better benchmark suit. 
```

Comment:
set up -> setup (No space in the noun.)
suit -> suite
_a_ better benchmark suite (but I'm actually not sure what's "bad" about Parsec)

Reply:
Our main issue with the benchmarks was getting them to run with our new pass. I'll modify the statement.